### PR TITLE
[PyTorch] Implement Fp8 padding and unpadding module

### PR DESCRIPTION
**Pull request by BeingGod**
# Description

Currently the FP8 unpadding backward is implemented by torch.autograd. it involves num_gemms * aten::fill + DtoD call that hurts the MoE model performance in FP8 training. So we implemented Fp8 padding and unpadding module to eliminate the overhead of autograd. The workflow shows below.

Workflow:
image

I show 2% E2E performance gain in our MoE model.

Fixes # (issue)

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refractor

## Changes

Please list the changes introduced in this PR:

- Implement a kernel `multi_padding_kernel` which fused multi-tensor padding.
- Implement `FP8Padding` and `FP8Unpadding` module.

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: transformer_engine/common/transpose/multi_pad_cast_transpose.cu
Comment (phu0ngng) [inline transformer_engine/common/transpose/multi_pad_cast_transpose.cu:1]: Hi,
I see quite a few code duplications between this file and the existing `cast_transpose.cu`. Is adding a padding option into `cast_transpose` and calling it from this `multi_pad_cast_transpose_kernel` possible?

Comment (BeingGod) [reply transformer_engine/common/transpose/multi_pad_cast_transpose.cu:1]: Hi @phu0ngng ,
I think `cast_transpose` is unnecessary to implement padding.
Reason: In MoE model routing will cause dimension of seq is not multiple of 16 so we should padding it. But for `cast_transpose` the most of dimension of seq are multiple of 16 (e.g 2048, 4096, 8192...).

**Comment by BeingGod** [comment]
I'm refactoring coding to remove implicit padding and unpadding operator.

Don't Merge.

File: transformer_engine/common/transpose/multi_pad_cast_transpose.cu
Comment (phu0ngng) [reply transformer_engine/common/transpose/multi_pad_cast_transpose.cu:1]: I mean we can avoid code duplications between `cast_transpose` and `multi_pad_cast_transpose_kernel` by templating `cast_transpose` so that it does padding when needed. Then from `multi_pad_cast_transpose_kernel`, one can iterate through the loop and call `cast_transpose` for each gemm with padding enabled.

Comment (BeingGod) [reply transformer_engine/common/transpose/multi_pad_cast_transpose.cu:1]: Ack. 

I'm refactoring code that `multi_pad_cast_transpose_kernel ` maybe removed. But your suggestion is valuable. Thanks a lot.

File: transformer_engine/pytorch/module/base.py [hunk 1]
```diff
@@ -659,6 +659,7 @@ def prepare_forward(
         is_first_microbatch: Union[bool, None],  # pylint: disable=unused-argument
         num_gemms: int = 1,
         allow_non_contiguous: bool = False,
+        with_param: bool = True,
```
Comment (yaox12) [inline transformer_engine/pytorch/module/base.py:662]: Seems never been used.

File: transformer_engine/pytorch/module/fp8_padding.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+        # Make sure input dimensions are compatible
+        in_features = inp.shape[-1]
+        inputmats = torch.split(inp.view(-1, in_features), m_splits)
+
+        # Allocate cast and transpose output tensor
+        total_row = sum(padded_m_splits)
+        out = torch.empty([total_row, in_features], dtype=inp.dtype, device="cuda")
+        out_list = torch.split(out, padded_m_splits)
+
+        multi_padding_fused(inputmats, padded_m_splits, out_list)
```
Comment (yaox12) [inline transformer_engine/pytorch/module/fp8_padding.py:42]: I think it's better to make `multi_padding_fused` accept whole tensors as inp/out, 
```python
inp = inp.view(-1, in_features)
multi_padding_fused(inp, m_splits, padded_m_splits, out)
```
and replace `torch.split` with stepping the pointers in C++ to reduce CPU overheads, as I'm doing in https://github.com/NVIDIA/TransformerEngine/pull/1128/files#diff-342b0e9e5b472b443484b3a2c4a78647cd72431c272ed44c678ecbe636fb7a3aR191-R194

**Review by yaox12 — COMMENTED** [review]
Generally LGTM.  
I feed we could have better names for `Fp8Padding` /`Fp8Unpadding`, such as `MultiPadding`/`MultiUnpadding`.  cc @phu0ngng

File: transformer_engine/pytorch/utils.py [hunk 3]
```diff
@@ -221,6 +221,14 @@ def cast_if_needed(tensor: torch.Tensor, dtype: torch.dtype) -> torch.Tensor:
         return tensor if tensor is None or tensor.dtype == dtype else tensor.to(dtype)
 
 
+def cast_if_needed_by_actual_dtype(
```
[the remark this answers is not in the document]
Comment (BeingGod) [reply transformer_engine/pytorch/utils.py:224]: Ack, thx.

File: transformer_engine/pytorch/module/base.py (hunk 1 above)
Comment (BeingGod) [reply transformer_engine/pytorch/module/base.py:662]: Ack thx.

File: transformer_engine/pytorch/module/fp8_padding.py (hunk 2 above)
Comment (BeingGod) [reply transformer_engine/pytorch/module/fp8_padding.py:42]: Ack. Your work is helpful to me. 

Thx.

**Comment by BeingGod** [comment]
> Generally LGTM. I feed we could have better names for `Fp8Padding` /`Fp8Unpadding`, such as `MultiPadding`/`MultiUnpadding`. cc @phu0ngng

Hi @yaox12, thanks for your suggestion. I wonder if `MultiPadding`/`MultiUnpadding` should be support customized padding number (e.g 8, 32...) ? 
Current padding and unpadding module only support padding to 16 (for FP8). So I added `FP8` prefix for padding and unpadding module.

**Comment by phu0ngng** [comment]
> Hi @yaox12, thanks for your suggestion. I wonder if `MultiPadding`/`MultiUnpadding` should support customized padding numbers (e.g. 8, 32...)? The current padding and unpadding module only supports padding to 16 (for FP8). So I added the `Fp8` prefix for padding and unpadding modules.

Are there any numbers on E2E performance gain in your MoE model for other non-FP8 types (BF16 for example) with this padding/unpadding?

**Comment by BeingGod** [comment]
> > Hi @yaox12, thanks for your suggestion. I wonder if `MultiPadding`/`MultiUnpadding` should support customized padding numbers (e.g. 8, 32...)? The current padding and unpadding module only supports padding to 16 (for FP8). So I added the `Fp8` prefix for padding and unpadding modules.
> 
> Are there any numbers on E2E performance gain in your MoE model for other non-FP8 types (BF16 for example) with this padding/unpadding?

I don't have non-FP8 types performance data now but it is a good idea. I will do some benchmark for non-FP8 types with padding/unpadding.

**Comment by phu0ngng** [comment]
> Hi @phu0ngng. I have done some benchmarks for BF16 with padding/unpadding. I set multiple of padding to 4,8,16. It seems to hurt E2E performance that uses padding/unpadding for BF16. Perhaps it is meaningless in the current that supports other multiple padding.

Agree. I think we can keep `FP8` in the name for now. Thanks.

File: tests/pytorch/test_numerics.py [hunk 4]
```diff
@@ -354,6 +357,40 @@ def forward(self, input: torch.Tensor) -> torch.Tensor:
         return (input > 0) * input * input
 
 
+class TorcGroupedLinearWithPadding(nn.Module):
```
Comment (phu0ngng) [inline tests/pytorch/test_numerics.py:360]: typo for `Torc` ?

File: tests/pytorch/test_numerics.py [hunk 5]
```diff
@@ -1328,6 +1365,158 @@ def test_grouped_linear_accuracy_parallel_mode(parallel_mode):
...
+        if p.requires_grad:
+            outputs.append(p.grad)
+    return outputs
+
+
+@pytest.mark.parametrize("dtype", param_types)
+@pytest.mark.parametrize("num_gemms", [3, 6])
+@pytest.mark.parametrize("bs", batch_sizes)
+@pytest.mark.parametrize("model", model_configs.keys())
+@pytest.mark.parametrize("fp8", all_boolean)
```
Comment (phu0ngng) [inline tests/pytorch/test_numerics.py:1464]: I think we can simply remove the `fp8 = false` case.

File: tests/pytorch/test_numerics.py [hunk 6]
```diff
@@ -354,6 +357,40 @@ def forward(self, input: torch.Tensor) -> torch.Tensor:
...
+        self.linear_fn = GroupedLinear(
+            num_gemms,
+            in_features,
+            out_features,
+            bias=bias,
+            params_dtype=params_dtype,
+            parallel_mode=parallel_mode,
+            device="cuda",
+        )
+        self.unpadding = Fp8Unpadding(num_gemms)
```
Comment (phu0ngng) [inline tests/pytorch/test_numerics.py:377]: A question here, this GroupedLinear is used for MoE only. If padding for FP8 increases 2% E2E performance, why don't we have this padding embedded in GroupedLinear by default?

File: transformer_engine/common/util/padding.cu [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+            const int n_blocks = kernel_args.block_range[kernel_args.num_tensors];
+            multi_padding_kernel<nvec_in, nvec_out, fp32, InputType, OutputType>
+            <<<n_blocks, threads_per_block, 0, stream>>>(kernel_args););  // NOLINT(*)
+    );                                                                    // NOLINT(*)
+  }
+}
+
+}  // namespace transformer_engine
+
+void nvte_multi_padding(size_t num_tensors, const NVTETensor* input_list, NVTETensor* output_list,
```
Comment (phu0ngng) [inline transformer_engine/common/util/padding.cu:221]: Hi, this kernel does both cast and padding. So I suggest renaming the kernel to make it more clear, for example `nvte_multi_cast_padding()`

File: transformer_engine/pytorch/csrc/extensions/padding.cu [hunk 8]
```diff
@@ -0,0 +1,79 @@
+/*************************************************************************
+ * Copyright (c) 2022-2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+ *
+ * See LICENSE for license information.
+ ************************************************************************/
+
+#include "extensions.h"
+
+void fused_multi_row_padding(at::Tensor input, at::Tensor output,
```
Comment (phu0ngng) [inline transformer_engine/pytorch/csrc/extensions/padding.cu:9]: Hi, why don't we simply pass a `std::vector<at::Tensor> input` so that we have all the size information and don't need to calculate the `d_input_ptr`, `d_output_ptr`?
This way was implemented in `te_grouped_gemm`.

File: tests/pytorch/test_numerics.py (hunk 4 above)
Comment (BeingGod) [reply tests/pytorch/test_numerics.py:360]: This is a `typo`. Thx.

File: tests/pytorch/test_numerics.py (hunk 5 above)
Comment (BeingGod) [reply tests/pytorch/test_numerics.py:1464]: Ack. Thx.

File: tests/pytorch/test_numerics.py (hunk 6 above)
Comment (BeingGod) [reply tests/pytorch/test_numerics.py:377]: In MLP Layer we have two successive GropedLinear if padding as a part of GroupedLinear we need paddding twice.

File: transformer_engine/pytorch/csrc/extensions/padding.cu (hunk 8 above)
Comment (BeingGod) [reply transformer_engine/pytorch/csrc/extensions/padding.cu:9]: The call of `aten::split` has about 30us CPU overhead. We replace it by calculating pointer address of device in plain C++ to reduce CPU overhead. 
Ref from @yaox12 : https://github.com/NVIDIA/TransformerEngine/pull/1128/files#diff-342b0e9e5b472b443484b3a2c4a78647cd72431c272ed44c678ecbe636fb7a3aR191-R194

File: transformer_engine/common/util/padding.cu (hunk 7 above)
Comment (BeingGod) [reply transformer_engine/common/util/padding.cu:221]: Ack.

Maybe I should simplify this kernel to remove cast. I hope this kernel can support all-type padding (including FP8). Its cast behavior is not fit to FP8.

File: tests/pytorch/test_numerics.py (hunk 6 above)
Comment (phu0ngng) [reply tests/pytorch/test_numerics.py:377]: Do we always call `moe_permutate` -> 2 x GroupedLinear -> `moe_unpermute`?

File: transformer_engine/pytorch/csrc/extensions/padding.cu (hunk 8 above)
Comment (phu0ngng) [reply transformer_engine/pytorch/csrc/extensions/padding.cu:9]: Ah, I missed that PR. Thanks

File: tests/pytorch/test_numerics.py (hunk 6 above)
Comment (BeingGod) [reply tests/pytorch/test_numerics.py:377]: Yes. In struct of transformer model MLP layer always has 2*Linear.

Comment (phu0ngng) [reply tests/pytorch/test_numerics.py:377]: Then, why don't we do FP8 padding in the permutation directly?

Comment (BeingGod) [reply tests/pytorch/test_numerics.py:377]: I already have evaluated this solution but in the permutation of MCore it have multi code path.
This is a good idea. Maybe we can try it later.

**Review by phu0ngng — APPROVED** [review]
LGTM!
